sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/parser/AstBuilder.scala

```diff
@@ -204,6 +207,41 @@ class AstBuilder extends SqlBaseBaseVisitor[AnyRef] with Logging {
   }
 
   /**
+   * Create a partition filter specification.
+   */
+  def visitPartitionFilterSpec(ctx: PartitionSpecContext): Expression = withOrigin(ctx) {
```

Comment:
Modify this. Let the AST builder make the expression(s) and validate them here. You can also introduce the AttributeReference here. For example:
```scala
val parts = ctx.partitionVal.asScala.map { pVal =>
  expression(pVal) match {
    case cmp @ BinaryComparison(UnresolvedAttribute(name), constant: Literal) =>
      cmp.withNewChildren(Seq(AttributeReference(name, StringType)(), constant))
    case _ => throw new ParseException(..., ctx)
  }
}

```